### PR TITLE
Fix: Inconsistent ports in canonical URLs

### LINKED ISSUE #198: Inconsistent ports in canonical URLs
When using the plugin on an archive like author or category on a domain like `http://localhost:10024/`, then the canonical URLs show the port twice. Example output:

```json
{"@context":"http:\/\/schema.org","@type":"WebPage","headline":"Author - admin","url":"http:\/\/localhost:10024:10024\/author\/admin\/"}	
```

On a single Post, then the port doesn't appear:

```json
{"@context":"http:\/\/schema.org","@type":"NewsArticle","mainEntityOfPage":{"@type":"WebPage","@id":"http:\/\/localhost\/post-2\/"},"headline":"Post 2","url":"http:\/\/localhost\/post-2\/","thumbnailUrl":"","image":{"@type":"ImageObject","url":""},"dateCreated":"2021-04-06T23:27:26Z","datePublished":"2021-04-06T23:27:26Z","dateModified":"2021-04-06T23:27:26Z","articleSection":"Uncategorized","author":[{"@type":"Person","name":"admin"}],"creator":["admin"],"publisher":{"@type":"Organization","name":"wp-parsely","logo":""},"keywords":[]}
```

On the homepage, it shows it correctly:

```json
{"@context":"http:\/\/schema.org","@type":"WebPage","headline":"wp-parsely","url":"http:\/\/localhost:10024"}
```

This suggests that the `get_current_url()` may need looking at, and tests added.

### PR DESCRIPTION
**Pull request by htdat**
Fix #198 

## Description 

This PR addresses inconsistent port issues for get_current_url() and a few other things: 

- Simplify the way we construct the current URL and respect the `force_https_canonicals` option. This is just a string replacement rather than take care of checking and concatenating scheme, host, URI, and port. 
- Change this function from `private` to `public` so that it can be tested with PHPUnit. 
- Add tests for this function.

Though this seems quite a lot of changes, I think it's worthwhile as this PR simplifies things and has tests. 

#### For the main issue #198 with the port number:

1. A simple way I found is to use this npm package https://developer.wordpress.org/block-editor/reference-guides/packages/packages-env/#quick-tldr-instructions
2. After that, visit the homepage, a post page, a few other pages (category, tag, author), and ensure that URLs are correct in the meta value `@context`.

#### For normal sites (without port number):

Basically, follow up step 2 above.

### CODE REVIEW

File: tests/all-test.php [hunk 1]
```diff
@@ -591,4 +591,87 @@ public function test_user_logged_in_multisite() {
...
+		$options                           = get_option( \Parsely::OPTIONS_KEY );
+		$options['force_https_canonicals'] = $force_https;
+		update_option( \Parsely::OPTIONS_KEY, $options );
+
+		update_option( 'siteurl', $url );
+
+		// Test homepage
+		$this->go_to( '/' );
+		$res = self::$parsely->get_current_url();
+		self::assertSame( strpos( $res, $expected, 0 ), 0 );
```
Comment (GaryJones) [inline tests/all-test.php]: Use a more specific assertion - https://phpunit.readthedocs.io/en/7.5/assertions.html#assertstringstartswith

File: tests/all-test.php [hunk 2]
```diff
@@ -591,4 +591,87 @@ public function test_user_logged_in_multisite() {
...
+
+		// Test a specific post
+		$post_array = $this->create_test_post_array();
+		$post       = $this->factory->post->create( $post_array );
+		$this->go_to( '/?p=' . $post );
+		$res = self::$parsely->get_current_url();
```
Comment (GaryJones) [inline tests/all-test.php]: You're testing against a post, but the default first argument of the method is `nonpost`. Seems like this should be `post`, along with the `$post` ID (rename variable as $post_id for clarity). Without `post` as the first arg value, there's a whole section of the `get_current_url()` method that isn't being covered in the tests.

Example:

```suggestion
		$post_id    = $this->factory->post->create( $post_array );
		$this->go_to( '/?p=' . $post_id );
		$res = self::$parsely->get_current_url( 'post', $post_id );
```

File: wp-parsely.php [hunk 3]
```diff
@@ -1649,29 +1648,19 @@ private function get_current_url( $post = 'nonpost', $post_id = 0 ) {
...
-				$canonical = $scheme . sanitize_text_field( wp_unslash( $_SERVER['HTTP_HOST'] ) ) . sanitize_text_field( wp_unslash( $_SERVER['REQUEST_URI'] ) ); // Input var okay.
-			}
+			$url = apply_filters( 'wp_parsely_permalink', $permalink, $post, $post_id );
+		} else {
+			$request_uri = isset( $_SERVER['REQUEST_URI'] )
+					? sanitize_text_field( wp_unslash( $_SERVER['REQUEST_URI'] ) )
+					: '';
 
-			return $canonical;
+			$url = site_url( $request_uri ); // Input var okay.
```
Comment (GaryJones) [inline wp-parsely.php]: That comment isn't doing anything.

File: wp-parsely.php [hunk 4]
```diff
@@ -1649,29 +1648,19 @@ private function get_current_url( $post = 'nonpost', $post_id = 0 ) {
...
 			 */
-			$permalink        = apply_filters( 'wp_parsely_permalink', $permalink, $post, $post_id );
-			$parsed_canonical = wp_parse_url( $permalink );
-			// handle issue if wp_parse_url doesn't return good host & path data, fallback to page url as a last resort.
-			if ( isset( $parsed_canonical['host'], $parsed_canonical['path'] ) ) {
-				$canonical = $scheme . $parsed_canonical['host'] . $parsed_canonical['path'];
-			} elseif ( isset( $_SERVER['HTTP_HOST'], $_SERVER['REQUEST_URI'] ) ) { // Input var okay.
-				$canonical = $scheme . sanitize_text_field( wp_unslash( $_SERVER['HTTP_HOST'] ) ) . sanitize_text_field( wp_unslash( $_SERVER['REQUEST_URI'] ) ); // Input var okay.
-			}
+			$url = apply_filters( 'wp_parsely_permalink', $permalink, $post, $post_id );
```
Comment (GaryJones) [inline wp-parsely.php:1649]: Can this filter be made available to the other branch as well? Does it make to do that?

File: wp-parsely.php [hunk 5]
```diff
@@ -1631,12 +1631,11 @@ private function get_settings_url() {
 	 * @param int    $post_id id of the post you want to get the url for. Optional.
 	 * @return string|void
 	 */
-	private function get_current_url( $post = 'nonpost', $post_id = 0 ) {
-		$options = $this->get_options();
-		$scheme  = ( $options['force_https_canonicals'] ? 'https://' : 'http://' );
+	public function get_current_url( $post = 'nonpost', $post_id = 0 ) {
+		$url = '';
```
Comment (GaryJones) [inline wp-parsely.php]: What does this assignment add?

File: tests/all-test.php [hunk 6]
```diff
@@ -591,4 +591,87 @@ public function test_user_logged_in_multisite() {
...
+			array(
+				false,
+				'http://example.com:1234',
+				'http://example.com:1234',
+			),
+			array(
+				false,
+				'https://example.com:1234',
+				'http://example.com:1234',
+			),
```
Comment (GaryJones) [inline tests/all-test.php]: How about mixing some of these up to test other strings, like `http://localhost:999` or `https://my-domain.local:1000/foo/bar`, since you're updating the site URL anyway?

**Review by GaryJones — CHANGES_REQUESTED** [review]

File: wp-parsely.php (hunk 4 above)
Comment (htdat) [reply wp-parsely.php:1649]: I keep it as-is like what the original code is doing.
I've looked at this filter again. It is passing `$post_id`, which is not available in another conditional branch so we should not make it available outside of the current scope.

File: wp-parsely.php (hunk 5 above)
Comment (htdat) [reply wp-parsely.php]: I do not have any strong opinion here. I just think it makes reading easier with a declaration somewhere. However, PHPStorm is also marking it's not necessary, then I am removing it.

File: wp-parsely.php [hunk 7]
```diff
@@ -1649,29 +1646,19 @@ private function get_current_url( $post = 'nonpost', $post_id = 0 ) {
...
+			$url = apply_filters( 'wp_parsely_permalink', $permalink, $post, $post_id );
+		} else {
+			$request_uri = isset( $_SERVER['REQUEST_URI'] )
+					? sanitize_text_field( wp_unslash( $_SERVER['REQUEST_URI'] ) )
+					: '';
 
-			return $canonical;
+			$url = home_url( $request_uri );
 		}
-		$page_url = site_url( null, $scheme );
```
Comment (htdat) [inline wp-parsely.php:1663]: Previously, this uses `site_url`. However, since we're dealing with the canonical links (front-end links), not WordPress file links, the correct usage should be `home_url`. 

I am also updating the relevant line in the test file to reflect this change too.

**Comment by htdat** [comment]
Thanks @GaryJones :D 
I've updated this PR to address your valuable feedback. In this iteration, the most major change is from `site_url` to `home_url` as I explained above.

**Comment by htdat** [comment]
I've just noticed we're having this issue too #174. It's directly related to this PR. 
I will continue researching and taking this issue into account.

**Review by GaryJones — APPROVED** [review]
Tests look good, changes look good, manual inspection of applied PR on local environment shows URLs displaying correctly 👍🏻
